chain/network/src/accounts_data/mod.rs

```diff
@@ -0,0 +1,206 @@
+//! Cache of AccountData. It keeps AccountData for important accounts for the current epoch.
+//! The set of important accounts for the given epoch is expected to never change (should be
+//! deterministic). Note that "important accounts for the current epoch" is not limited to
+//! "validators of the current epoch", but rather may include for example "validators of the next
+//! epoch" so that AccountData of future validators is broadcasted in advance.
+//!
+//! Assumptions:
+//! - verifying signatures is expensive, we need a dedicated threadpool for handling that.
+//!   TODO(gprusak): it would be nice to have a benchmark for that
+//! - a bad peer may attack by sending a lot of invalid signatures
+//! - we can afford verifying each valid signature of the current epoch once.
+//! - we can afford verifying a few invalid signatures per SyncAccountsData message.
+//!
+//! Strategy:
+//! - handling of SyncAccountsData should be throttled by PeerActor/PeerManagerActor.
+//! - synchronously select interesting AccountData (i.e. those with never timestamp than any
+//!   previously seen for the given (account_id,epoch_id) pair.
+//! - asynchronously verify signatures, until an invalid signature is encountered.
+//! - if any signature is invalid, drop validation of the remaining signature and ban the peer
+//! - all valid signatures verified, so far should be inserted, since otherwise we are open to the
+//!   following attack:
+//!     - a bad peer may spam us with <N valid AccountData> + <1 invalid AccountData>
+//!     - we would validate everything every time, realizing that the last one is invalid, then
+//!       discarding the progress
+//!     - banning a peer wouldn't help since peers are anonymous, so a single attacker can act as a
+//!       lot of peers
+use crate::network_protocol;
+use crate::network_protocol::SignedAccountData;
+use near_network_primitives::types::AccountKeys;
+use near_primitives::types::{AccountId, EpochId};
+use parking_lot::RwLock;
+use rayon::iter::{ParallelBridge, ParallelIterator};
+use std::collections::HashMap;
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::Arc;
+
+#[cfg(test)]
+mod tests;
+
+async fn rayon_spawn<T: 'static + Send>(f: impl 'static + Send + FnOnce() -> T) -> T {
+    let (send, recv) = tokio::sync::oneshot::channel();
+    rayon::spawn(move || send.send(f()).ok().unwrap());
```

Comment:
Hm, I'd `let _` instead of `unwrap` here. Futures are cancellable, and it's hard to rule out the case where something somewhere above the stack ends up in a `select` or `timeout`. 

To be more specific: I think putting a future which is not explicitly prepated for cancellation into select is the mistake, but its an easy mistake to make, so I'd love us to sort-of defensively code in this situation: just wasting the time computing `f` seems better than crashing the whole process.

Reply:
Added a must_complete wrapper for futures which are expected to complete.
If an incomplete future is dropped, the wrapper calls log_assert (which panics in tests, and logs an error in prod). Also added a log_assert wrapping the send.send().is_ok() call.